In [ ]:
!pip install spacy==3.3.0

# Processamento de Linguagem Natural aplicada à Gestão Pública

Aula 2 (01/06): Ferramentas de processamento de texto 

# Configuração do Colab

Importar pacotes, carregar dados, aplicar definições.

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [ ]:
import spacy
#import spacy.cli

#spacy.cli.download("pt_core_news_md")
#nlp = spacy.load("pt_core_news_md")
#nlp.to_disk("/content/drive/MyDrive/oficina_nlp/models")
nlp = spacy.load("/content/drive/MyDrive/oficina_nlp/models")

## Exercício 01 - Municípios mineiros

In [ ]:
import pandas as pd

In [ ]:
transferencias_municipais = pd.read_excel("/content/drive/My Drive/oficina_nlp/data_raw/transf_municipios.xlsx")
transferencias_municipais

nome          icms        ipi          ipva  \
0       ABADIA DOS DOURADOS  1.778.837,63  23.964,07    476.921,83   
1                    ABAETE  3.277.622,25  43.903,34  1.675.294,71   
2                ABRE CAMPO  1.730.765,27  23.482,30  1.041.798,12   
3                   ACAIACA    694.879,27   9.212,26    168.008,78   
4                   ACUCENA  1.217.106,01  16.349,29    271.443,57   
..                      ...           ...        ...           ...   
848            VIRGINOPOLIS  1.434.494,90  19.228,18    648.458,21   
849             VIRGOLANDIA    764.306,74  10.233,43    127.982,02   
850  VISCONDE DO RIO BRANCO  6.247.824,70  83.365,22  3.197.892,67   
851            VOLTA GRANDE  1.411.199,98  19.058,07    203.971,28   
852          WENCESLAU BRAZ    597.202,43   8.059,50     94.167,29   

            total  
0    2.279.723,53  
1    4.996.820,30  
2    2.796.045,69  
3      872.100,31  
4    1.504.898,87  
..            ...  
848  2.102.181,29  
849    902.522,19  
850  9.529.082,59  
851  1.634.229,33  
852    699.429,22  

[853 rows x 5 columns]

In [ ]:
populacao_estimada = pd.read_excel("/content/drive/My Drive/oficina_nlp/data_raw/populacao_estimada_ibge.xlsx")
populacao_estimada.dropna(inplace=True)
populacao_estimada

cod_ibge              municipio populacao_estimada
0     1100015  Alta Floresta D'Oeste              22516
1     1100023              Ariquemes             111148
2     1100031                 Cabixi               5067
3     1100049                 Cacoal              86416
4     1100056             Cerejeiras              16088
...       ...                    ...                ...
5565  5222005             Vianópolis              14088
5566  5222054         Vicentinópolis               9002
5567  5222203               Vila Boa               6451
5568  5222302          Vila Propício               5941
5569  5300108               Brasília            3094325

[5570 rows x 3 columns]

In [ ]:
uf = populacao_estimada["cod_ibge"].str.slice(0,2)
uf

0       11
1       11
2       11
3       11
4       11
        ..
5565    52
5566    52
5567    52
5568    52
5569    53
Name: cod_ibge, Length: 5570, dtype: object

In [ ]:
populacao_estimada = populacao_estimada[uf == "31"]
populacao_estimada

cod_ibge               municipio populacao_estimada
2244  3100104     Abadia dos Dourados               7022
2245  3100203                  Abaeté              23263
2246  3100302              Abre Campo              13434
2247  3100401                 Acaiaca               3995
2248  3100500                 Açucena               9270
...       ...                     ...                ...
3092  3171808            Virginópolis              10459
3093  3171907             Virgolândia               5303
3094  3172004  Visconde do Rio Branco              43351
3095  3172103            Volta Grande               5270
3096  3172202          Wenceslau Braz               2543

[853 rows x 3 columns]

Escreva, no bloco de código a seguir, uma função que padroniza um dado texto aplicando duas operações:

1.   Converte todos os caracteres em maiúsculo
2.   Converte todos os caracteres para a codificação ASCII



In [ ]:
def padroniza_texto(texto):
  # Escreva seu código aqui
  
  

Aplique a função criada na coluna `municipio` da tabela `populacao_estimada`:

In [ ]:
populacao_estimada["municipio_padr"] = populacao_estimada["municipio"].apply(_________)
populacao_estimada

Faça o cruzamento ("merge" ou "join") das tabelas:

In [ ]:
populacao_estimada.merge(transferencias_municipais, left_on="municipio_padr", right_on="________")

cod_ibge               municipio populacao_estimada  \
0    3100104     Abadia dos Dourados               7022   
1    3100203                  Abaeté              23263   
2    3100302              Abre Campo              13434   
3    3100401                 Acaiaca               3995   
4    3100500                 Açucena               9270   
..       ...                     ...                ...   
839  3171808            Virginópolis              10459   
840  3171907             Virgolândia               5303   
841  3172004  Visconde do Rio Branco              43351   
842  3172103            Volta Grande               5270   
843  3172202          Wenceslau Braz               2543   

             municipio_padr                    nome          icms        ipi  \
0       ABADIA DOS DOURADOS     ABADIA DOS DOURADOS  1.778.837,63  23.964,07   
1                    ABAETE                  ABAETE  3.277.622,25  43.903,34   
2                ABRE CAMPO              ABRE CAMPO  1.730.765,27  23.482,30   
3                   ACAIACA                 ACAIACA    694.879,27   9.212,26   
4                   ACUCENA                 ACUCENA  1.217.106,01  16.349,29   
..                      ...                     ...           ...        ...   
839            VIRGINOPOLIS            VIRGINOPOLIS  1.434.494,90  19.228,18   
840             VIRGOLANDIA             VIRGOLANDIA    764.306,74  10.233,43   
841  VISCONDE DO RIO BRANCO  VISCONDE DO RIO BRANCO  6.247.824,70  83.365,22   
842            VOLTA GRANDE            VOLTA GRANDE  1.411.199,98  19.058,07   
843          WENCESLAU BRAZ          WENCESLAU BRAZ    597.202,43   8.059,50   

             ipva         total  
0      476.921,83  2.279.723,53  
1    1.675.294,71  4.996.820,30  
2    1.041.798,12  2.796.045,69  
3      168.008,78    872.100,31  
4      271.443,57  1.504.898,87  
..            ...           ...  
839    648.458,21  2.102.181,29  
840    127.982,02    902.522,19  
841  3.197.892,67  9.529.082,59  
842    203.971,28  1.634.229,33  
843     94.167,29    699.429,22  

[844 rows x 9 columns]

Observem que a tabela cruzada contém apenas 844 registros. Alguns municípios não foram cruzados.

Podemos verificar para quais municípios não houve correspondência através do objeto *set* (conjunto), para o qual existem métodos como a diferença e a diferença simétrica.

Exemplo:

In [1]:
A = {1, 2, 3}
B = {2, 3, 4}

print(type(A))

<class 'set'>


In [ ]:
A.

In [3]:
A.intersection(B)

{2, 3}

In [4]:
A.union(B)

{1, 2, 3, 4}

In [5]:
A.difference(B)

{1}

In [6]:
B.difference(A)

{4}

In [7]:
A.symmetric_difference(B)

{1, 4}

Utilize a função `set` para converter as colunas dos nomes de municípios de cada tabela em conjuntos e aplique o método apropriado para identificar os nomes sem correspondência.

In [ ]:
A = set(populacao_estimada["__________"])
B = set(transferencias_municipais["_________"])

sem_corresp = A.________(B)
print(sem_corresp)

Complete o código a seguir para exibir os nomes na tabela de população que não constam na tabela de transferências:

In [ ]:
U = _.difference(_)
print("Nomes na tabela de população que não constam na tabela de transferências:")
print(U)

Complete o código a seguir para exibir os nomes na tabela de transferências que não constam na tabela de população:

In [ ]:
V = _.difference(_)
print("Nomes na tabela de população que não constam na tabela de transferências:")
print(V)

Vamos verificar as grafias de um mesmo município nas tabelas:

In [ ]:
populacao_estimada[populacao_estimada["municipio_padr"].str.contains("AMPARO DO SERRA")]

cod_ibge        municipio populacao_estimada   municipio_padr
2270  3102506  Amparo do Serra               4643  AMPARO DO SERRA

In [ ]:
transferencias_municipais[transferencias_municipais["nome"].str.contains("AMPARO DA SERRA")]

nome        icms        ipi        ipva         total
27  AMPARO DA SERRA  906.570,42  12.265,71  164.350,01  1.083.186,14

Por fim, altere, na base das transferências municipais, os nomes dos municípios sem correspondência.

In [ ]:
# Insira o seu código aqui




In [ ]:
# Verificando se temos 100% de correspondência
set(populacao_estimada["municipio_padr"]).symmetric_difference(set(transferencias_municipais["nome"]))

Revisão:


*   Seu código é *reutilizável*? (Caso o conjunto de palavras sem correspondência fosse outro, o mesmo código funcionaria?)

Vimos que alguns municípios estão com pequenas diferenças na redação do nome. **Poderíamos fazer um código que retorna o cruzamento dos municípios aceitando pequenas diferenças na escrita? Como?**